File: solutions/0543-diameter-of-binary-tree/solution.js
var diameterOfBinaryTree = function(root) {
    let maxDiameter = 0;  // Track maximum diameter found
    
    function dfs(node) {
        // Base case: null node has height -1 to make leaf nodes have diameter 0
        if (node === null) return -1;
        
        // Get heights of left and right subtrees
        const leftHeight = dfs(node.left);
        const rightHeight = dfs(node.right);
        
        // Update max diameter if current node's path is longer
        // Diameter at current node = leftHeight + rightHeight + 2
        maxDiameter = Math.max(maxDiameter, leftHeight + rightHeight + 2);
        
        // Return height of current node
        return Math.max(leftHeight, rightHeight) + 1;
    }
    
    dfs(root);
    return maxDiameter;
};

File: solutions/0543-diameter-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_diameter = 0
        
        def dfs(node):
            if not node:
                return -1
            
            left_height = dfs(node.left)
            right_height = dfs(node.right)

            # because left and right is guaranteed to have at least 2 here
            self.max_diameter = max(self.max_diameter, left_height + right_height + 2)
            return max(left_height, right_height) + 1
        
        dfs(root)
        return self.max_diameter
